### PR TITLE
Add damped newton solver, tidy solution model processing

### PR DESCRIPTION
**Pull request by bobmyhill**
The PR does three things in preparation for an equilibrium gibbs calculator:
- Add a derivative-based damped affine invariant newton solver (Deuflhard, 1974;1975;2004) with optional constraints. I have added several standard tests for the solver.
- Neaten up the solution model chemical processing. This is mostly cosmetic, although I have added a new return attributes (site_names) and function (solution_bounds), which will make gibbs calculations more straightforward.
- Add a __str__ function to burnman.composite.Composite for pretty printing purposes.

### CODE REVIEW

File: tests/test_solvers.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import unittest
```
Comment (tjhei) [inline tests/test_solvers.py]: is this executed or do we need to list it in ``tests.py``?

File: burnman/composite.py [hunk 2]
```diff
@@ -72,6 +73,21 @@ def __init__(self, phases, fractions=None, fraction_type='molar'):
             self.molar_fractions = None
 
         self.set_averaging_scheme('VoigtReussHill')
+        self.name=name
+        
+    def __str__(self):
```
Comment (tjhei) [inline burnman/composite.py]: This is unrelated to the PR. I assume you want it for the tests? We have ``to_string()`` in Material, Composite, etc.  which is not very pythonic but and this is very similar to ``debug_print`` (did you see this function?). The ``self.name`` should probably appear in ``to_string()`` and/or ``debug_print`` if you insist on keeping things this way.

File: burnman/nonlinear_solvers.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    with Jacobian J(x) and constraints C(x), using the 
+    damped affine invariant modification to Newton's method 
+    (Deuflhard, 1974;1975;2004), 
+    Here we follow the algorithm as described in Nowak and Weimann (1991):
+    [Technical Report TR-91-10, Algorithm B]
+
+    The iteration continues until three conditions are satisfied:
+    all(np.abs(dx (simplified newton step) < tol))
+    all(np.abs(dx (full Newton step) < sqrt(10*tol))) [avoiding pathological behaviour] and
+    lambda = lambda_bounds(dx, x)[1] (lambda = 1 for a full Newton step).
```
Comment (tjhei) [inline burnman/nonlinear_solvers.py]: are you missing a comparison here (``lambda<0`` or something?)?

File: burnman/nonlinear_solvers.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+    Inequality constraints are provided by the function C(x), which returns a
```
Comment (tjhei) [inline burnman/nonlinear_solvers.py]: there is no ``C`` argument

File: burnman/nonlinear_solvers.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        Maximum number of iterations for solver.
+    lambda_bounds: function
+        Returns a tuple of floats (1.e-8, 1.) corresponding
+        to the minimum and maximum allowed fractions of the
+        full newton step (dx). A function of dx and x.
+    constraints : function
+        Returns the LHS of the inequality constraints(x)
+        as a 1D numpy array. The constraints are satisfied if
+        all the elements of the array are less than or equal to
+        zero.
```
Comment (tjhei) [inline burnman/nonlinear_solvers.py]: "A function of x"

File: burnman/composite.py (hunk 2 above)
Comment (bobmyhill) [reply burnman/composite.py]: Yes, I should probably have put this in a separate PR. 
I think it might be worthwhile replacing `to_string()` with `__str__()` in all classes. That way, when anyone wants to see key information about an object, they can just type `print(object)`, rather than `print(object.to_string())`. What do you think?

**Comment by bobmyhill** [comment]
matplotlib was also giving me a bunch of deprecation warnings, so I modernised the plotting commands in a couple of examples.

File: burnman/composite.py (hunk 2 above)
Comment (bobmyhill) [reply burnman/composite.py]: For the time being, I've just added self.name to to_string and debug_print.
